docs/nsm_api.md

```diff
@@ -0,0 +1,445 @@
+Network Service Mesh (NSM) API
+
+## Executive summary
+
+Based on the discussions on NSM IRC channel on September 11th/12th and in the issue [https://github.com/ligato/networkservicemesh/issues/283](https://github.com/ligato/networkservicemesh/issues/283). This document describes API calls between different components in NSM enabled kubernetes cluster. Each type of API is provided with proto file definition and brief description of parameters expected and returned. Developers of NFVi who plan to leverage NSM for their applications, are urged to thoroughly review these API calls to make sure that expected and returned parameters do fully cover their applications&#39; needs.
+
+
+
+## List of Network Service Mesh Components
+
+### Network Service Mesh Components in the Abstract
+
+Current Network Service Mesh is highly focused on Network Service Mesh within the context of a Kubernetes Cluster.  Most of this document will work within that context.
+Network Service Mesh concepts are highly genericizable.
+
+![Network Service Mesh Abstract Concepts](./images/NSM%20Diagrams%20for%20Arch%20Docs.png)
+<dl>
+    <dt>Network Service (NS)</dt>
+    <dd>
+        A Network Service is the abstract representation of the various behaviors to be provided to a Network Service Client via an L2/L3 connection.  It can include:
+        <ul>
+            <li>Connectivity to Isolated Resources</li>
+            <li>Protection for Threats</li>
+            <li>Guaranteed Bandwidth/Latency</li>
+            <li>Load Balancing</li>
+            <li>Proxying</li>
+        </ul>
+        A Network Service has a name and a payload.  The payload defines the type of L2/L3 payload (Ethernet/IPv4/IPv6/MPLS/etc) the Network Service accepts.
+        Examples:
+        <ul>
+            <li>Secure Intranet Access</li>
+            <li>Bride Domains</li>
+            <li>Routing Domains (VRFs)</li>
+            <li>Cloud-native Network Functions (CNFs) of all sorts</li>
+        </ul>
+    </dd>
+    <dt>Network Service Client (NSC)</dt>
+    <dd>
+        A Network Service Client is an entity which wishes to connect to a Network Service.  
+        Example: A Pod in Kubernetes which wants to connect to a Network Service.
+    </dd>
+    <dt>L2/L3 Connection</dt>
+    <dd>
+        An L2/L3 Connection is simply some mechanism which can carry L2/L3 traffic (IP/Ethernet/MPLS/etc) bidirectionally between the Network Service Client and the Network Service Endpoint.  L2/L3 Connections are simple cross connects.  They are not bridge domains.  A bridge domain is itself a Network Service, and can be connected to via an L2/L3 Connection.
+    </dd>
+    <dt>Network Service Endpoint (NSE)</dt>
+    <dd>
+        A Network Service Endpoint provides a concrete usable instance of a Network Service for consumption by Network Service Clients.  Example: Pods providing a Network Service in Kubernetes.
+    </dd>
+</dl>
```

Comment:
I think the current model is to have  a common DaemonSet (Network Service Manager) for all data planes and then have a NSE and NSMD per data plane. This would have (potentially) multiple NSE Pods running one for each MSMD and it would need one per Node?. Would an alternative model would have a unique DaemonSet per NSMD that was deployed on every node that needed the NSM Type. This would allow the  Kubernetes scheduler put client Pods on Nodes with the appropriate DS. It would make the DS more complex and tied to the NSMD but would remove an additional Pod that needs to be scheduled to a specific Node - I think...